internal/client.go

```diff
@@ -16,152 +16,207 @@ import (
+func NewClient(opts ...ClientOption) (*Client, error) {
+	c := &Client{
+		httpClient: http.DefaultClient,
+	}
+	for _, opt := range opts {
+		if err := opt(c); err != nil {
+			return nil, err
 		}
+	}
 
-		resp, err := baseClient.client.Do(req)
+	// If the user does not specify a logger, then we'll use zap for a default one
+	if c.logger == nil {
+		zapLog, err := zap.NewProduction()
 		if err != nil {
-			return baseClient, err
+			return nil, err
 		}
-		defer func() {
-			if rerr := resp.Body.Close(); err == nil {
-				err = rerr
-			}
-		}()
-
-		err = CheckResponse(resp)
-
-		baseClient.Cookies = resp.Cookies()
+		logger := zapr.NewLogger(zapLog)
+		c.logger = &logger
 	}
 
-	return baseClient, nil
-}
-
-// NewBaseClient returns a basic http/https client based on isHttp flag
-func NewBaseClient(credentials *prismgoclient.Credentials, absolutePath string, isHTTP bool) (*Client, error) {
-	if absolutePath == "" {
+	if c.UserAgent == "" {
+		return nil, fmt.Errorf("userAgent argument must be passed")
+	}
+	if c.absolutePath == "" {
 		return nil, fmt.Errorf("absolutePath argument must be passed")
 	}
-
-	httpClient := http.DefaultClient
+	if c.credentials == nil {
+		return nil, fmt.Errorf("credentials argument must be passed")
+	}
+	if c.BaseURL == nil {
+		c.logger.V(1).Info("BaseURL is not set. Using URL from credentials", "url", c.credentials.URL)
+		if err := WithBaseURL(c.credentials.URL)(c); err != nil {
+			return nil, err
+		}
+	}
 
 	transCfg := &http.Transport{
 		// to skip/unskip SSL certificate validation
 		TLSClientConfig: &tls.Config{
-			InsecureSkipVerify: credentials.Insecure,
+			InsecureSkipVerify: c.credentials.Insecure,
 		},
 	}
-	httpClient.Transport = logging.NewTransport("Nutanix", transCfg)
-
-	protocol := httpsPrefix
-	if isHTTP {
-		protocol = httpPrefix
+	if c.credentials.ProxyURL != "" {
+		c.logger.V(1).Info("Using proxy:", "proxy", c.credentials.ProxyURL)
+		proxy, err := url.Parse(c.credentials.ProxyURL)
+		if err != nil {
+			return nil, fmt.Errorf("error parsing proxy url: %s", err)
+		}
+		transCfg.Proxy = http.ProxyURL(proxy)
 	}
+	c.httpClient.Transport = transCfg
 
-	baseURL, err := url.Parse(fmt.Sprintf(defaultBaseURL, protocol, credentials.URL))
-	if err != nil {
-		return nil, err
-	}
+	if c.credentials.SessionAuth {
+		c.logger.V(1).Info("Using session_auth")
+
+		ctx := context.TODO()
+		req, err := c.NewRequest(ctx, http.MethodGet, "/users/me", nil)
+		if err != nil {
+			return nil, err
+		}
+
+		resp, err := c.httpClient.Do(req)
+		if err != nil {
+			return nil, err
+		}
+		defer resp.Body.Close()
 
-	c := &Client{credentials, httpClient, baseURL, "", nil, nil, absolutePath, ""}
+		if err := CheckResponse(resp); err != nil {
+			return nil, err
+		}
+
+		c.cookies = resp.Cookies()
+	}
 
 	return c, nil
 }
 
 // NewRequest creates a request
 func (c *Client) NewRequest(ctx context.Context, method, urlStr string, body interface{}) (*http.Request, error) {
```

Comment:
Is context used here at all?
We should allocate http request using this context to propagate timeout.
See https://pkg.go.dev/net/http#NewRequestWithContext

Reply:
This is just request creation, there's no need for passing context in the parameter; will remove it altogether.